### PR TITLE
fix(types): keras+tf refactor LazyLoader with type-safe

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

## Description

A few things tackle in this PR
- depends on #2054 for stubs
- Move keras and tensorflow into `_internal.frameworks`, then lazy import it in `__init__.py`
- adds stubs for type hint since static type checker won't be able to know the type
image
image

## Checklist:

- [ ] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/keras.py [hunk 1]
```diff
@@ -187,39 +176,59 @@ def __init__(
...
-        with self._device:
+        raw_predict_fn = getattr(self._model, self._predict_fn_name)
+        self._predict_fn = functools.partial(raw_predict_fn, **self._partial_kwargs)
+
+    def _run_batch(
+        self,
+        *args: t.Union[
+            t.List[t.Union[int, float]], "np.ndarray[t.Any, np.dtype[t.Any]]", tf.Tensor
+        ],
+        **kwargs: t.Any,
```
Comment (larme) [inline bentoml/keras.py:198]: @aarnphm  does keras really need this? The `predict` should only accept one input (multiple inputs are inside a list). So we may just need partial kwargs. And here partial kwargs is not flexible like in pytorch's case, just some configuration parameter for keras's predict function. So in my pr I renamed it into predict_kwargs (#2069)

Comment (aarnphm) [reply bentoml/keras.py]: noted. Was trying to make pyright happy :)

File: tests/integration/frameworks/test_keras_impl.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def test_keras_save_load(model, kwargs, modelstore):
     if not TF2:
         # Initialize variables in the graph/model
         bentoml.keras._sess.run(tf.global_variables_initializer())
-        with bentoml.keras._default_sess():
+        with bentoml.keras._sess.as_default():
```
Comment (parano) [inline tests/integration/frameworks/test_keras_impl.py:57]: is this line required for our users as well?

Comment (parano) [reply tests/integration/frameworks/test_keras_impl.py:57]: it would be great to add some basic sample usage to the docstring

Comment (aarnphm) [reply tests/integration/frameworks/test_keras_impl.py:57]: for tf1 case then yes. i will add some docstring for that one.

Comment (aarnphm) [reply tests/integration/frameworks/test_keras_impl.py:69]: this can be accessed via `get_session()`

**Comment by bojiang** [comment]
_Had a discussion with Aaron on Slack._

* keep using LazyLoader combine with `if TYPE_CHECKING`. it's the best solution so far, before a typed LazyLoader implementation with pyright plugin
* `overload` is not a better `Union`/`Optional`. Use it only when the possible combination of params are less than cartesian product

File: bentoml/__init__.py [hunk 3]
```diff
@@ -13,11 +15,18 @@
 from ._internal.service import Service
 from ._internal.service.loader import load
 from ._internal.types import Tag
+from ._internal.utils import LazyLoader as _LazyLoader
 
 # bento APIs are top-level
 from .bentos import list  # pylint: disable=W0622
 from .bentos import build, delete, export_bento, get, import_bento
 
+if TYPE_CHECKING:
```
Comment (aarnphm) [inline bentoml/__init__.py:24]: @parano

File: bentoml/_internal/utils/tensorflow.py [hunk 4]
```diff
@@ -1,14 +1,17 @@
-import logging
+import importlib.util
```
Comment (aarnphm) [inline bentoml/_internal/utils/tensorflow.py:1]: Also I should fix this, we shouldn't rely on this https://github.com/asottile/scratch/wiki/ImportSideEffects

Comment (aarnphm) [reply bentoml/_internal/utils/tensorflow.py:1]: However, autoflake `--remove-unused-imports` seems to rely on this behaviour

Comment (aarnphm) [reply bentoml/_internal/utils/tensorflow.py:1]: we don't use autoflake but there might be scenario where developers use it. Put it here for future reference

**Review by parano — APPROVED** [review]

File: bentoml/_internal/utils/tensorflow.py (hunk 4 above)
Comment (parano) [reply bentoml/_internal/utils/tensorflow.py:1]: does it help if we move this to the get_tf_version function below?

**Pull request merged**